### PR TITLE
tools/vms-generator: use runStrategy for example VMs instead of running

### PR DESCRIPTION
**Pull request by dasionov**
### What this PR does
Before this PR:
example vms are being generated with running field

After this PR:
examples vms will be generated with runStrategy instead of running

relates to issue: #11993
Partially addresses issue #12285 by replacing `spec.running` with `spec.runStrategy`

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place:

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
```release-note
NONE
```

### CODE REVIEW

**Review by EdDev — COMMENTED** [review]
I am guess this change is related to the intention to deprecate the `running` field with the `runStrategy` one.

I can only guess this is only one small portion of that effort, but it still needs a reference to the reasoning and the discussions held on the matter.
Then, to provide a convincing commit/PR comment that explains why it is advised to switch from one to the other.

I do think there is a need to communicate this decition and the actions that are planned first with Kubevirt community, assuring awareness and providing a chance for feedback.

---

Please try to follow the commit text recommendations as specified [here](https://cbea.ms/git-commit/#seven-rules).
I also like to have the subject line prefixed with a title: `<title>: <subject>`.
E.g. `tools, vms-generator: Replace 'running' with 'runStrategy' in examples`

**Comment by EdDev** [comment]
> update vms-generator to use runStrategy instead of running

IMO this is not a user-facing change that merits a release-note.

File: tools/vms-generator/utils/utils.go
Comment (alicefr) [inline tools/vms-generator/utils/utils.go:702]: Can we use the pointer kubevirt package here?
```suggestion
			RunStrategy: pointer.P(v1.RunStrategyHalted),
```

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @dasionov.
Could you please add reasoning to the commit message about why is this change needed?

File: tools/vms-generator/utils/utils.go
Comment (dasionov) [reply tools/vms-generator/utils/utils.go:644]: do you think i should replace usages of `k6tpointer "kubevirt.io/kubevirt/pkg/pointer"` and `"k8s.io/utils/pointer"` to be used with `ptr.To` everywhere in the file ? or it's out of the scope

Comment (codingben) [reply tools/vms-generator/utils/utils.go:644]: I think it would be out of the scope, would be awesome to see it in the following PR.

Comment (dankenigsberg) [reply tools/vms-generator/utils/utils.go:644]: @EdDev has a strong opinion about avoiding external dependencies, when we have simple implementation inside. Please keep `k6tpointer` unless you have a strong reason not to.

Comment (codingben) [reply tools/vms-generator/utils/utils.go:644]: @EdDev Why it's needed to have pointer package where in k8s's ptr package you have the exact same implementation? Shouldn't we get rid of this package?

```
// pkg/pointer
func P[T any](t T) *T {
	return &t
}

// k8s.io/utils/ptr
func To[T any](v T) *T {
	return &v
}
```

By the way, what does `k6t` mean in `k6tpointer`? Can you just name it `pointer`?

Comment (orelmisan) [reply tools/vms-generator/utils/utils.go:644]: > By the way, what does k6t mean in k6tpointer? Can you just name it pointer?

k6t is an abbreviation for KubeVirt like K8s is an abbreviation for Kubernetes.

**Comment by codingben** [comment]
@EdDev Why it's needed to have pointer package where in k8s's ptr package you have the exact same implementation? Shouldn't we get rid of this package?

```
// pkg/pointer
func P[T any](t T) *T {
	return &t
}

// k8s.io/utils/ptr
func To[T any](v T) *T {
	return &v
}
```

By the way, what does `k6t` mean in `k6tpointer`? Can you just name it `pointer`?

**Comment by dasionov** [comment]
> @EdDev Why it's needed to have pointer package where in k8s's ptr package you have the exact same implementation? Shouldn't we get rid of this package?
> 
> ```
> // pkg/pointer
> func P[T any](t T) *T {
> 	return &t
> }
> 
> // k8s.io/utils/ptr
> func To[T any](v T) *T {
> 	return &v
> }
> ```
> 
> By the way, what does `k6t` mean in `k6tpointer`? Can you just name it `pointer`?

Was asking myself the same question, but from Dan answer I realized the main reason is to be able to use pointer without importing k8s dependencies

**Review by dankenigsberg — APPROVED** [review]

File: tools/vms-generator/utils/utils.go
Comment (enp0s3) [reply tools/vms-generator/utils/utils.go:644]: @codingben Hi, we always strive to reduce dependency on external packages unless its really needed, especially those that introduce small self-contained toolbox for your project. Main reason is that its easier to maintain, you don't to be dependent on external project policies.

**Comment by orelmisan** [comment]
@dasionov Please consider stating in the PR description that this PR partially addresses #12285.

**Comment by dasionov** [comment]
> @dasionov Please consider stating in the PR description that this PR partially addresses #12285.

Thank you for pointing this one out :+1:

**Comment by EdDev** [comment]
> but I also would like to know if there are other reasons for us to have this package.

It is a combination of history, consistency and reasonable use of external dependency.
Originally, we used an older bigger package from k8s utils, then generics came but the alternative was not released there due to usages from other projects. At that time, it made sense to just write 3 lines and drop the dependency, or not add one when writing new code.

Today, you’ll need to convince me why would we be interested to depend on something external for just 3 lines. I see no added value.
If it was some logic that relates to k8s or is non trivial, I would probably think differently.

File: tools/vms-generator/utils/utils.go
Comment (iholder101) [reply tools/vms-generator/utils/utils.go:644]: > @EdDev has a strong opinion about avoiding external dependencies, when we have simple implementation inside. Please keep `k6tpointer` unless you have a strong reason not to.

+1 to not adding dependencies. I'm with @EdDev on this one.

> Why it's needed to have pointer package where in k8s's ptr package you have the exact same implementation?

Just because the implementation is a one-liner (or 3? :smile:) it does not worth bringing in another dependency IMO. Also, in this case we prefer to avoid third part updates to this package to keep it as simple as it is now, unless there's a strong reason to do otherwise.

**Comment by iholder101** [comment]
> I do think there is a need to communicate this decition and the actions that are planned first with Kubevirt community, assuring awareness and providing a chance for feedback.

Thanks @EdDev! I completely agree.
I've opened this issue https://github.com/kubevirt/kubevirt/issues/11993. As expressed there, I think the next steps would be to discuss it in both our mailing list and a design proposal. This wasn't done mainly due to the fact that I was out-of-office for a while, and due to the fact that we're currently discussing an [alternative for design proposals](https://github.com/kubevirt/community/pull/288), but I'll get to it soon! Thanks for the reminder.

**Comment by iholder101** [comment]
BTW @dasionov, a small tip for the future: I think it's easier to review if the auto-generated code resides in its own commit. In other word, a separate commit for the `make generate`. This PR is small enough so it doesn't matter a lot, but for large amounts of code I think it would make a difference :)

Thanks a lot for this PR!

**Comment by dasionov** [comment]
> BTW @dasionov, a small tip for the future: I think it's easier to review if the auto-generated code resides in its own commit. In other word, a separate commit for the `make generate`. This PR is small enough so it doesn't matter a lot, but for large amounts of code I think it would make a difference :)
> 
> Thanks a lot for this PR!

Thanks for the tip, I completely agree with you

File: tools/vms-generator/utils/utils.go
Comment (codingben) [reply tools/vms-generator/utils/utils.go:644]: @enp0s3 @iholder101 Thanks for replying. I respect you but I disagree with you guys. It's just my opinion and you can approve it without my approval of course, I'm not even an approver or reviewer here.

----

Since we're using Kubernetes objects, to me it also makes sense to use any other utils provded by Kubernetes library.

I see that `pointer` package in KubeVirt was introduced in Dec 2022 with main one reason, according to Eddy [1]:
```
It is aimed to replace `k8s.io/utils/pointer` with a generic version.
Unlike the previous version, the generic one accepts any type.
```
Because before you had this:
```
pointer.Bool(secureBoot)
```
And with Go 1.18 that added generics support you can do:
```
pointer.P(secureBoot)
```

IMO, this is the real reason that `pointer` package is there until today. IMO, it's no longer needed because it can be replaced by `k8s.io/utils/ptr` that supports generic types.

[1] https://github.com/kubevirt/kubevirt/commit/7b7a6fd86dbe57bbadcb09d25bdbf3be1d3291cf

Comment (codingben) [reply tools/vms-generator/utils/utils.go:644]: Just a final note: This dependency `k8s.io/utils/pointer` is already used widely in KubeVirt codebase. Someone needs to convince contributors not to use `k8s.io/utils/ptr` instead in the future.

Comment (iholder101) [reply tools/vms-generator/utils/utils.go:644]: > @enp0s3 @iholder101 Thanks for replying. I respect you but I disagree with you guys. It's just my opinion and you can approve it without my approval of course, I'm not even an approver or reviewer here.

> to me it also makes sense to use any other utils provded by Kubernetes library.

Thanks for the discussion @codingben!

It takes courage to share your opinion as an underdog. I deeply encourage you to always share your honest opinion, as it's perfectly okay (and even beneficial!) to disagree :)

It makes sense to rely on external dependencies sometimes, and we definitely do that in Kubevirt. However, there's a cost for external dependencies that we have to consider, and there's always a trade-off.

I respect you and your opinion, but in the end of the day we have to make a choice. In this specific case there seems to be a consensus around not bringing in this external dependency. Since I saw there's a large consensus for it, I've approved the PR.

Thanks again for this discussion and for your honest opinion!

Comment (iholder101) [reply tools/vms-generator/utils/utils.go:644]: > Just a final note: This dependency `k8s.io/utils/pointer` is already used widely in KubeVirt codebase, it's deprecated and suggested by Kubernetes maintainers to be replaced by `k8s.io/utils/ptr`.

You're right. These usages predate the current Kubevirt pointer package, and we should definitely drop this dependency moving forward.

Comment (codingben) [reply tools/vms-generator/utils/utils.go:644]: > It takes courage to share your opinion as an underdog.

I disagree. It's just another request that came from OpenShift Virtualization management, especially from @dankenigsberg and @EdDev to review code cleanup PRs in kubevirt/kubevirt repository. From the beginning it felt wrong - exactly like "underdog".

cc @dominikholler

Comment (iholder101) [reply tools/vms-generator/utils/utils.go:644]: By "underdog" I meant that most of the people that took part in the discussion had the other opinion.
As I've said above, I deeply encourage you to always share your honest opinion even if there's disagreement.

**Pull request merged**
